### DISCUSSION TITLE
Refactor Carrier to allow more & clean customizations in modules

### DISCUSSION THREAD

**Discussion by rmilecki** [PrestaShop/PrestaShop#33405] [Ideas]
#### Is your feature request related to a problem?

PrestaShop built-in carrier supports basic shipping cost rules (based on country and either: order total or order weight):
![ranges](https://user-images.githubusercontent.com/452972/93721573-7f223780-fb91-11ea-9d50-1518f8a980cf.png)
Some shops may need more complex rules (or calculations) than that. E.g.:

1. Shipping price based on both: order total *and* order weight
2. Shipping price based on product types (e.g. extra fee if there is one or more food product)
3. Shipping price based on amount of parcels needed
4. Shipping method not available if there isn't any oversize product
5. Shipping method available only for specific city / cities

Long store short: there are always going to be some corner cases requiring custom modules, Carrier-s and settings / calculations. Ideally PrestaShop should make it possible & simple to build modules providing custom Carrier-s.

#### Describe the solution you'd like

I think that `CarrierCore` should:

1. Have basic entry points `checkDeliveryPrice()` and `getMaxDeliveryPrice()` that would internally check for 1 of supported shipping methods:
- `Carrier::SHIPPING_METHOD_WEIGHT`
- `Carrier::SHIPPING_METHOD_PRICE`
- `Carrier::SHIPPING_METHOD_FREE`
- `Carrier::SHIPPING_METHOD_CUSTOM` (new value)
2. Execute single hook with cart (and address?) as argument(s). We may consider passing also a shipping code based on weight / price ranges. That could allow custom modules to do some custom calculation on top of basic range tables.
3. Allow hook to return e.g. `false` in case shipping method is unavailable for a given cart (products) / address / totals / whatever.

#### Alternatives you've considered

`Carrier` executes hooks `actionDeliveryPriceByWeight` and `actionDeliveryPriceByPrice` but they don't allow enough customizations:

1. They still require selecting between "According to total price." and "According to total weight." even if they ignore whole ranges table afterwards.
2. They can't access list of cart products (they only get `total_weight` or `order_total`)
3. They can't report carrier as unavailable (for whatever reason)

The whole usage of above hooks is a bit misleading. Functions in `Carrier` first calculate shipping cost and then execute hooks that can totally overwrite the previous shipping cost result.

#### Additional context

This request may be considered while working on the #10586
Request that could make use of this one: #15379

[a comment by an automated account is not shown]

**Comment by sarahdib** [comment]
@PrestaShop/product-team Could you check this improvement please ?
Thank you!

**Comment by LouiseBonnard** [comment]
Hi @rmilecki, thanks for this idea! I've just added it to the related EPIC issue (#19690) so that it can be studied and addressed along with other problems/improvements when this page will be reworked.

As for the hooks, I add @PrestaShop/prestashop-core-developers in the loop: what would be your recommendation regarding this idea?

**Comment by rmilecki** [comment]
During the last 2 weeks I learned that:
1. Custom carrier module can set `need_range` to `1` in case it doesn't use "According to total price." / "According to total weight.". That is totally ignored by `AdminCarrierWizard` though.
Moreover it seem like an alternative to the `Carrier::SHIPPING_METHOD_WEIGHT` and `Carrier::SHIPPING_METHOD_PRICE` so it should be rather something like `Carrier::SHIPPING_METHOD_CUSTOM` instead of separated property (and column table).
2. Custom carrier modules **can** implement custom shipping cost calculation method.
It's just an undocumented & hidden feature. For details you have to see `Cart::getPackageShippingCostFromModule` (yes, `Cart`!). It can be done by implementing one of following methods:
   - `getOrderShippingCostExternal()`
   - `getPackageShippingCost()`
   - `getOrderShippingCost()`

I think current code state is a bit messy and I don't know how to clean it up properly and I'm not sure if that can be done without breaking backward compatibility. It also doesn't help that above solutions were added without any explanation. See:
- 42b4711504a6 ("// BugFix with carrier module and multishipping")
- 13a86f2dcab3 ("Add hooks actionDeliveryPriceByPrice and actionDeliveryPriceByWeight to customize delivery prices")

**Comment by kpodemski** [comment]
@MatShir something to think about while analyzing Shipping pages :)
